python/lsst/ts/Dome/dome_csc.py

```diff
@@ -598,21 +600,37 @@ async def implement_simulation_mode(self, simulation_mode):
                 f"Simulation_mode={simulation_mode} must be 0 or 1"
             )
 
-    # noinspection PyMethodMayBeStatic
-    async def schedule_task_periodically(self, period, task):
-        """Schedules a task periodically.
-
-        Parameters
-        ----------
-        period : `float`
-            The period in (decimal) seconds at which to schedule the function.
-        task : coroutine
-            The function to be scheduled periodically.
+    async def schedule_llc_tasks_periodically(self):
+        """Schedules the LLC status tasks periodically.
         """
+        start_tai = salobj.current_tai()
         try:
             while True:
-                await task()
-                await asyncio.sleep(period)
+                current_tai = salobj.current_tai()
+                reset_start_tai = False
+                if (current_tai - start_tai) > _AMCS_PERIOD:
+                    await self.statusAMCS()
+                    reset_start_tai = True
+                if (current_tai - start_tai) > _APsCS_PERIOD:
+                    await self.statusApSCS()
+                    reset_start_tai = True
+                if (current_tai - start_tai) > _LCS_PERIOD:
+                    await self.statusLCS()
+                    reset_start_tai = True
+                if (current_tai - start_tai) > _LWSCS_PERIOD:
+                    await self.statusLWSCS()
+                    reset_start_tai = True
+                if (current_tai - start_tai) > _MonCS_PERIOD:
+                    await self.statusMonCS()
+                    reset_start_tai = True
+                if (current_tai - start_tai) > _ThCS_PERIOD:
+                    await self.statusThCS()
+                    reset_start_tai = True
+
+                if reset_start_tai:
+                    start_tai = current_tai
```

Comment:
I strongly suspect this code is broken. The problem is that you reset start_tai if *any* status is output, which means that any status with a longer period will never run.

You could keep a separate start time for each method, but that's a lot of state to keep track of and a lot of repetition. Also this design means you can't set truly independent intervals for the various kinds of status.

Consider running a separate loop for each status. If you hold them in a collection then they are easy to cancel when shutting down or go out of ENABLED/DISABLED.
```
def __init__(self, ...):
    self.status_tasks = []

async def one_status_loop(self, method, interval):
    """Run one status method forever at the specified interval.
    ...
    """
    try:
        while True:
            method()
            await asyncio.sleep(interval)
    except Exception:
        self.log.exception(f"one_status_loop({method}) failed")

async def cancel_status_tasks(self):
    "Cancel all status tasks."""
    while self.status_tasks:
        self.status_tasks.pop().cancel()

async def start_status_tasks(self):
    """Start all status tasks
    """
    self.cancel_status_tasks()
    for method, interval in ((self.statusAMCS, AMCS_STATUS_INTERVAL)), ...):
        self.status_tasks.append(asyncio.create_task(self.one_status_loop(method, interval))
```
You could also make start_status_tasks wait for all the tasks to finish, in which case you could just cancel the task that runs it to shut down all status. But you have to be a bit careful in your use of `gather` so that if one status task fails it doesn't leave the main task dead and a bunch of subtasks running (makes for a lot of warnings at shutdown). But I think the above is a bit simpler, and it assures that all status tasks are canceled.

Reply:
Thanks very much for the suggestion. I adopted it and it works like a charm.